NewRelic/AutoInteractor.php

```diff
@@ -0,0 +1,162 @@
+<?php
+
+/*
+ * This file is part of Ekino New Relic bundle.
+ *
+ * (c) Ekino - Thomas Rabaix <thomas.rabaix@ekino.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Ekino\Bundle\NewRelicBundle\NewRelic;
+
+/**
+ * This interactor does never assume that the NewRelic extension is installed. It will check
+ * for the existence of each method EVERY time. This is a good interactor to use when you want
+ * to enable and disable the NewRelic extension without rebuilding your container.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class AutoInteractor implements NewRelicInteractorInterface
```

Comment:
What about this? It will make sure that (A) it's testable and (B) the code only has to check if new relic is present during the initialization.

```php
{
    private $interactor;
    private $loaded;

    public function __construct(NewRelicInteractorInterface $interactor, $forceLoading = false)
    {
        $this->interactor = $interactor;
        $this->loaded = $forceLoading || extension_loaded('newrelic');
    }

    /**
     * {@inheritdoc}
     */
    public function setApplicationName($name, $key = null, $xmit = false)
    {
        $this->loaded && $this->interactor->setApplicationName($name, $key, $xmit);
    }

    /**
     * {@inheritdoc}
     */
    public function setTransactionName($name)
    {
        $this->loaded && $this->interactor->setTransactionName($name);
    }

    /**
     * {@inheritdoc}
     */
    public function ignoreTransaction()
    {
        $this->loaded && $this->interactor->ignoreTransaction();
    }

    /**
     * {@inheritdoc}
     */
    public function addCustomMetric($name, $value)
    {
        $this->loaded && $this->interactor->addCustomMetric($name, $value);
    }

    /**
     * {@inheritdoc}
     */
    public function addCustomParameter($name, $value)
    {
        $this->loaded && $this->interactor->addCustomParameter($name, $value);
    }

    /**
     * {@inheritdoc}
     */
    public function getBrowserTimingHeader()
    {
        if (!$this->loaded) {
            return '';
        }

        return $this->interactor->getBrowserTimingHeader();
    }

    /**
     * {@inheritdoc}
     */
    public function getBrowserTimingFooter()
    {
        if (!$this->loaded) {
            return '';
        }

        return $this->interactor->getBrowserTimingFooter();
    }

    /**
     * {@inheritdoc}
     */
    public function disableAutoRUM()
    {
        $this->loaded && $this->interactor->disableAutoRUM();
    }

    /**
     * {@inheritdoc}
     */
    public function noticeError($msg)
    {
        $this->loaded && $this->interactor->noticeError($msg);
    }

    /**
     * {@inheritdoc}
     */
    public function noticeException(\Exception $e)
    {
        $this->loaded && $this->interactor->noticeException($e);
    }

    /**
     * {@inheritdoc}
     */
    public function enableBackgroundJob()
    {
        $this->loaded && $this->interactor->enableBackgroundJob();
    }

    /**
     * {@inheritdoc}
     */
    public function disableBackgroundJob()
    {
        $this->loaded && $this->interactor->disableBackgroundJob();
    }

    /**
     * {@inheritdoc}
     */
    public function startTransaction($name)
    {
        $this->loaded && $this->interactor->startTransaction($name);
    }

    /**
     * {@inheritdoc}
     */
    public function endTransaction()
    {
        $this->loaded && $this->interactor->endTransaction();
    }
}
```